### ISSUE TITLE
result structure gets incomplete support

### ISSUE THREAD

**Issue by ghost** [SoarGroup/Soar#139]
_From [marin...@gmail.com](https://code.google.com/u/110912605412702728125/) on May 29, 2013 14:15:28_

Attached is a simple Soar program that drops into a substate and then returns a result structure (^foo.bar ran) to the superstate.

There are two substate apply rules (with the same name). The first one is a standard apply rule that directly creates the result structure. This works as expected -- the substate goes away and the result structure remains. (To see this, reorder the applies so this one comes second, or comment out the other one -- they have the same name.)

The second one tries to be clever and uses a description of the result structure on the operator to create the result. It does this by firing twice in parallel. In this case, when the substate goes away, the ^foo part stays, but the ^bar ran part retracts. Looking at the justifications, only the ^foo part is included in the justification.

I suspect that, because the rule fires twice in parallel, that the instantiation that creates ^bar ran is local to the substate, and thus only o-supported wrt the substate. The ^foo part is a proper result that just happens to have ^bar ran as substructure (but that substructure was not actually returned as a result). So Soar does not recognize ^bar ran as a result, and it retracts with the substate.

Laird comments:

I took a quick look. This looks to be a race condition, where it creates (F1 ^bar ran) in parallel with the creation of (S1 ^foo F1). When these two are created in the same rule, Soar is smart enough to look for the connection between them, but since this is happening in rules firing in parallel, it is not smart enough to examine all of the parallel actions (that create other preferences) and notice that F1 is now in the closure of S1, so (F1 ^bar ran) should be a result. As you can imagine, doing this right would require it to do some extra processing (multiple passes through all of the preferences to chain through dependencies) that I'm sure Soar isn't doing.  So, yes it is a bug, ....

**Attachment:** [test.soar](http://code.google.com/p/soar/issues/detail?id=139)

_Original issue: http://code.google.com/p/soar/issues/detail?id=139_

**Comment by bluechill** [comment]
@marinier was this ever fixed? If not can you find test.soar again?

**Comment by marinier** [comment]
Not fixed to my knowledge. Here is test.soar.

**Comment by mazina** [comment]
I'm trying to see if this is still a problem, but I'm not sure what the correct behavior is that you expect.  Can you tell me what to look for, Bob?

**Comment by mazina** [comment]
I don't think this is really a bug, though it is an interesting case.  

The underlying problem is that chunking is determining the results based on what's in working memory when the base instantiation that started learning fires.  Neither of those working memory elements exist at the time the chunk is formed.  They're still pending preferences.  And I don't think we could use them to figure it that (F1 ^bar ran)  should be a result either, since other preferences that may not exist yet can play a role.  In other words, we can't detect that (F1 ^bar ran) has become accessible to the superstate, i.e. is a result, until after preference resolution happens in a later phase.  

I think this is another manifestation of the general problem of augmenting intermediate results.  If your second rule fired a cycle later rather than simultaneously, it would have still failed because it wouldn't be able to find a path to ground (F1 ^bar ran) to (S1 ^foo F1), since the latter was never directly tested in the substate by a rule, since it was created by a rule in the substate.

This is an interesting case, because EBC can now repair a rule that augments previous results and return a good chunk, but it still can't repair the simultaneous case.  I do have an idea for repairing learned rules by tracking substate symbols that get promoted and using conditions from the rule that promoted them to fix the ungrounded rule.  I think that would actually handle the simultaneous case, as long as I assume the promotions are guaranteed to happen as soon as it's detected.
